languages/elixir/exercises/concept/tuples/test/kitchen_calculator_test.exs

```diff
@@ -0,0 +1,89 @@
+defmodule KitchenCalculatorTest do
+  use ExUnit.Case
+
+  alias KitchenCalculator, as: KC
+
+  describe "get volume from tuple pair" do
+    test "get cups" do
+      assert KC.get_volume({:cup, 1}) == 1
+    end
+
+    test "get fluid ounces" do
+      assert KC.get_volume({:fluid_ounce, 2}) == 2
+    end
+
+    test "get teaspoons" do
+      assert KC.get_volume({:teaspoons, 3}) == 3
+    end
+
+    test "get tablespoons" do
+      assert KC.get_volume({:tablespoons, 4}) == 4
+    end
+
+    test "get millilitres" do
+      assert KC.get_volume({:millilitre, 5}) == 5
+    end
+  end
+
+  describe "convert to millilitres from" do
+    test "millilitres" do
+      assert KC.to_millilitre({:millilitre, 3}) == {:millilitre, 3}
+    end
+
+    test "cups" do
+      assert KC.to_millilitre({:cup, 3}) == {:millilitre, 720}
+    end
+
+    test "fluid ounces" do
+      assert KC.to_millilitre({:fluid_ounce, 100}) == {:millilitre, 3000}
+    end
+
+    test "teaspoon" do
+      assert KC.to_millilitre({:teaspoon, 3}) == {:millilitre, 15}
+    end
+
+    test "tablespoon" do
+      assert KC.to_millilitre({:tablespoon, 3}) == {:millilitre, 45}
+    end
+  end
+
+  describe "convert from millilitres to" do
+    test "millilitres" do
+      assert KC.from_millilitre({:millilitre, 4}, :millilitre) == {:millilitre, 4}
+    end
+
+    test "cups" do
+      assert KC.from_millilitre({:millilitre, 960}, :cup) == {:cup, 4}
+    end
+
+    test "fluid ounces" do
+      assert KC.from_millilitre({:millilitre, 6000}, :fluid_ounce) == {:fluid_ounce, 200}
+    end
+
+    test "teaspoon" do
+      assert KC.from_millilitre({:millilitre, 60}, :teaspoon) == {:teaspoon, 12}
+    end
+
+    test "tablespoon" do
+      assert KC.from_millilitre({:millilitre, 120}, :tablespoon) == {:tablespoon, 8}
```

Comment:
All of these test cases use numbers that produce nice integer results, which means we don't check what the students do with the decimal part in case it exist. Is that on purpose?

Reply:
The test cases are on purpose.  As you said above, I haven't taken time in this exercise to _teach_ floating point numbers, but I think the purpose of the exercise being to focus on tuples/pattern matching so I didn't want to distract with rounding errors but floating point numbers are probably the correct type for math involving measurements.

I could alter some of the test cases to end in low decimal place answers if you think that might help implicitly clear things up for students.

Reply:
> I haven't taken time in this exercise to teach floating point numbers, but I think the purpose of the exercise being to focus on tuples/pattern matching so I didn't want to distract with rounding errors but floating point numbers are probably the correct type for math involving measurements.

The design document states:

> - Know how to write a floating point literal
> - Know how to round a floating point number to decimal place or integer

It's absolutely fine to divert from that. You've used nice integers in the tests that when divided, end up with whole numbers. I think that is a very sensible approach here if you don't want to introduce floating-point numbers (which makes sense). So maybe just removed the floating-point bit from the design document?

Reply:
> I think the purpose of the exercise being to focus on tuples/pattern matching

Totally! In that case, what about writing the whole thing with integers-only (using integers in all of the examples introduction, instructions, hints) and pretending like the problem of floats doesn't exist? I agree that floats are more reasonable for math involving measurements, but in this exercise they seem to be a distraction.

Btw. I don't know all of the details. Is it technically possible for division of obviously divisible integers, like `4 / 2`, to also have rounding errors? E.g. `4 / 2` returns `2.0000000001` and thus the assertion `2.0000000001 == 2` fails?

Reply:
I added some decimal cases in the revision which result in low precision decimal numbers.

Interesting question, and yes, some whole numbers are not representable with floating point numbers in Elixir (which are always double precision), this is an interesting SO discussion about this here: https://stackoverflow.com/questions/19473352/are-there-any-whole-numbers-which-the-double-cannot-represent-within-the-min-max

About division, I am not sure of any obvious cases, but with addition, there is the famous case of `0.1 + 0.2 = 0.30000000000000004`